### ISSUE TITLE
`crate = {}` is accepted as a dependency constraint

### ISSUE THREAD

**Issue by huonw** [rust-lang/cargo#2147]
```rust
[package]
authors = []
name = "test"
version = "0.1.0"

[dependencies]
crossbeam = {}
```

`cargo build` runs fine, downloading the latest crossbeam. (It seems to act somewhat like `"*"`.)

cc @alexcrichton

**Comment by Zarathustra30** [comment]
Is this actually a misfeature? `{}` says "I don't care about this" just as well as `"*"` does.

**Comment by huonw** [comment]
I disagree with "just as well". `"*"` is at least an explicit version constraint, even if it is very loose, hence implies that it will be downloaded from crates.io (etc.), while `{}` isn't being explicit about anything.

**Comment by alexcrichton** [comment]
Yeah this is a misfeature, Cargo should give an error about this.

This will probably have to be a warning for at least one cycle, however, to mitigate any breakage.

**Comment by felixc** [comment]
I've got a patch ready that emits a warning under these conditions, but I have two questions before I send a PR:

1) Is there a convention for the phrasing of a warning like this that indicates possible future breakage? At the moment I have "`warning: Dependency (name-of-dep) specified without providing a local path, Git repository, or version to use.`", but that doesn't really let the user know that this may break in a version or two.

2) Is there some marker or annotation I should leave in the code (or other such process) to indicate that this should be cleaned up and made into an error in the future? Or do we simply leave this bug open and hope that in a release or two someone gets around to looking at it and cleaning it up?

**Comment by felixc** [comment]
Actually, those seem like good questions to discuss in a PR :)

Sent #2270

**Comment by felixc** [comment]
Fixed in #2270

**Comment by huonw** [comment]
Woohoo!
